File: src/main/java/me/rabbittv/terraTrees/utils/AlignTree.java
package me.rabbittv.terraTrees.utils;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;

public class AlignTree {

    public static Location alignSouth(Location origin, Material saplingType) {
        World world = origin.getWorld();
        if (world == null) return null;

        int[][] offsets = {
                {0, 0}, {-1, 0}, {0, -1}, {-1, -1}
        };

        for (int[] offset : offsets) {
            int dx = offset[0];
            int dz = offset[1];

            Location sw = origin.clone().add(dx, 0, dz);
            Location se = sw.clone().add(1, 0, 0);
            Location nw = sw.clone().add(0, 0, 1);
            Location ne = sw.clone().add(1, 0, 1);

            if (world.getBlockAt(sw).getType() == saplingType &&
                    world.getBlockAt(se).getType() == saplingType &&
                    world.getBlockAt(nw).getType() == saplingType &&
                    world.getBlockAt(ne).getType() == saplingType) {
                return sw;
            }
        }

        return null;
    }

}

File: src/main/kotlin/me/rabbittv/terraTrees/utils/AlignTree.kt
package me.rabbittv.terraTrees.utils

import org.bukkit.Material
import org.bukkit.Location

class AlignTree {

    fun alignSouth(origin: Location, saplingType: Material): Location? {
        val world = origin.getWorld() ?: return null

        val offsets = arrayOf(
            intArrayOf(0, 0),
            intArrayOf(-1, 0),
            intArrayOf(0, -1),
            intArrayOf(-1, -1)
        )
        for (offset in offsets) {
            val dx = offset[0]
            val dz = offset[1]
            val dx_double = dx.toDouble()
            val dz_double = dz.toDouble()

            val sw = origin.clone().add(dx_double, 0.0, dz_double)
            val se = sw.clone().add(1.0, 0.0, 0.0)
            val nw = sw.clone().add(0.0, 0.0, 1.0)
            val ne = se.clone().add(1.0, 0.0, 1.0)

            if (world.getBlockAt(sw).type == saplingType &&
                world.getBlockAt(se).type == saplingType &&
                world.getBlockAt(nw).type == saplingType &&
                world.getBlockAt(ne).type == saplingType
            ) {
                return sw
            }
        }
        return null
    }

}